### PR TITLE
My grade function final

### LINKED ISSUE #23: raneem(my_grade)
""" A function that gives your grade after adding your GPA"""
so it is a function where you add your GPA and you will be getting your grade (A, B..etc) along with its description (example: excellent, good... etc).

### PR DESCRIPTION
**Pull request by Raneemrami**
---
name: my_grade
about: It should evaluate your 4.0 scale GPA to your final grade (A, B, B+ ..etc) with its description.  
---

It is a function where you add your GPA and get your final alphabetical grade along with its description.
Examples:
It evaluates 3.6 to "Excellent (A)"
It evaluates 2.99 to "Good (B)"
It evaluates 1.6 to "Basic (D)"
Issue link:
[#23](https://github.com/MIT-Emerging-Talent/ET6-foundations-group-31/issues/23)

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Comment by Raneemrami** [comment]
this fixes #23

File: solutions/my_grade.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+@author: Raneem Rami
+"""
+
+
+def my_grade(average: float) -> str:
+    """Gives the grade (A,B,B+,C,D,F) and a grade description for a 4.0 scale GPA.
+    Parameters:
+        average: float, your GPA to get it evaluated to your grade, must not be greater than 4.0
+                and must not be negative.
+
```
Comment (MayMon-T3807) [inline solutions/my_grade.py:18]: I think parameter documentation needs improvement. like adding specific  specific return value documentation like

"""
Returns:
    str: A string containing the grade description and letter grade in parentheses.
         Format: "{Description} ({Letter})"
         Possible values: "Excellent (A)", "Very Good (B+)", "Good (B)", 
                         "Satisfactory (C)", "Basic (D)", "Fail (F)"
"""

File: solutions/tests/test_my_grade.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        """It should evaluate 3.5 to "Very good (B+)" """
+        self.assertEqual(my_grade(3.5), "Very Good (B+)")
+
+    # Test for Good grade (B)
+    def test_good(self):
+        """It should evaluate 2.8 to "Good (B)" """
+        self.assertEqual(my_grade(2.8), "Good (B)")
+
+    # Test for Satisfactory grade (C)
+    def test_satisfactory(self):
```
Comment (MayMon-T3807) [inline solutions/tests/test_my_grade.py:35]: Test docstrings should be more consistent like satisfactory' is lowercase in docstring but capitalized in assertion

**Review by MayMon-T3807 — CHANGES_REQUESTED** [review]
Overall, the code provides a functional GPA to letter grade converter with good test coverage. However, I think there are several areas that need improvement:

Changes Requested:

1. **Naming**:
   - Rename `my_grade` to something more descriptive like `gpa_to_letter_grade`
   - Current name doesn't clearly indicate the function's purpose

2. **Documentation**:
   - Enhance parameter documentation with clearer GPA range specifications
   - Make return value documentation more explicit about output format 

3. **Tests**:
   - Add missing edge case tests (e.g., 2.99, 3.59)
   - Fix inconsistent capitalization in test docstrings
   - Fix the unittest main block syntax (`**name**` -> `__name__`)

4. **Implementation**:
   - Consider using constants for grade boundaries
   - Remove redundant comments that just repeat the code
   - Rename 'average' parameter to 'gpa' for clarity

File: solutions/tests/test_my_grade.py (hunk 2 above)
Comment (Raneemrami) [reply solutions/tests/test_my_grade.py:35]: wait, I will check

Comment (Raneemrami) [reply solutions/tests/test_my_grade.py:35]: sorry May, what are the tests that need fix can you help me detect them

File: solutions/my_grade.py (hunk 1 above)
Comment (Raneemrami) [reply solutions/my_grade.py:18]: I added that, can you check?

File: solutions/tests/test_my_grade.py (hunk 2 above)
Comment (MayMon-T3807) [reply solutions/tests/test_my_grade.py:35]: okay like using uppercase and lower case letter to be more consistent. like
# Test for satisfactory grade (C)
    def test_satisfactory(self):

Comment (Raneemrami) [reply solutions/tests/test_my_grade.py:35]: so you want me to add the alphabetical grade along with its description in the test's name. right?

Comment (Raneemrami) [reply solutions/tests/test_my_grade.py:35]: is there any other problem?

File: solutions/my_grade.py (hunk 1 above)
Comment (Raneemrami) [reply solutions/my_grade.py:18]: wait I am fixing

**Review by MayMon-T3807 — APPROVED** [review]
that's look better now.

File: solutions/tests/test_my_grade.py (hunk 2 above)
Comment (MayMon-T3807) [reply solutions/tests/test_my_grade.py:35]: yes
no problem if you don't add

File: solutions/my_grade.py (hunk 1 above)
Comment (MyatCharm) [reply solutions/my_grade.py:18]: Awesome. Maymon's feedback is constructive. I agree with all of the recommendations. Did 'Edge Case' test has been added, sorry if I missed.

**Comment by MyatCharm** [comment]
I suppose it looks good now. I'll merge now
